vms/components/fee/dimensions.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fee
+
+import "github.com/ava-labs/avalanchego/utils/math"
+
+const (
+	Bandwidth Dimension = iota
+	DBRead
+	DBWrite // includes deletes
+	Compute
+
+	NumDimensions = iota
+)
+
+type (
+	Dimension  uint
+	Dimensions [NumDimensions]uint64
+)
+
+func (d Dimensions) Add(o Dimensions) (Dimensions, error) {
+	var (
+		res Dimensions
+		err error
+	)
+	for i := range d {
+		res[i], err = math.Add64(d[i], o[i])
+		if err != nil {
+			return res, err
+		}
+	}
+	return res, nil
+}
+
+func (d Dimensions) Sub(o Dimensions) (Dimensions, error) {
+	var (
+		res Dimensions
+		err error
+	)
+	for i := range d {
+		res[i], err = math.Sub(d[i], o[i])
+		if err != nil {
+			return res, err
+		}
+	}
+	return res, nil
+}
+
+func (d Dimensions) ToGas(weights Dimensions) (Gas, error) {
+	var res uint64
+	for i := range d {
+		v, err := math.Mul64(d[i], weights[i])
+		if err != nil {
+			return 0, err
+		}
+		res, err = math.Add64(res, v)
+		if err != nil {
+			return 0, err
+		}
+	}
+	return Gas(res), nil
```

Comment:
I didn't understand why it was divided by 10 to begin with... It isn't referenced by the ACP: https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/103-dynamic-fees#dimensions

Reply:
You're right it was not. This came out while tuning the parameters.
Essentially if dimension weights are very differents (e.g. Size and DBRead/Write around 10 units, Compute around 1000 units), you may want to choose weights that normalize them to weight around the same (so 100s for Size/DBs 1 for Compute).
This may increase gas too much, so you may want to scale it back, say by 10.
The division may cause numerical errors while building the tx.
I suggest we update code and proposal to convert dimensions to gas with scalar product + denominator to renomalize.

Reply:
I think for now I'd prefer to keep the code clean. If the spec changes then we can update as we go imo